ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -211,25 +216,25 @@ openDBImpl :: (HasCallStack, IOLike m, Ord blockId, NoUnexpectedThunks blockId)
            -> ThrowCantCatch (VolatileDBError blockId) (STM m)
            -> Parser e m blockId
            -> Int -- ^ @maxBlocksPerFile@
+           -> ResourceRegistry m
            -> m (VolatileDBEnv m blockId)
-openDBImpl hasFS@HasFS{..} err errSTM parser maxBlocksPerFile =
+openDBImpl hasFS@HasFS{..} err errSTM parser maxBlocksPerFile registry =
     if maxBlocksPerFile <= 0
     then EH.throwError err $ UserError . InvalidArgumentsError $
       "maxBlocksPerFile should be positive"
     else do
-      st <- mkInternalStateDB hasFS err parser maxBlocksPerFile
+      st <- mkInternalStateDB hasFS err parser maxBlocksPerFile registry
       stVar <- newMVar $ VolatileDbOpen st
-      return $ VolatileDBEnv hasFS err errSTM stVar maxBlocksPerFile parser
+      return $
+        VolatileDBEnv hasFS err errSTM stVar maxBlocksPerFile parser registry
 
 closeDBImpl :: IOLike m
             => VolatileDBEnv m blockId
             -> m ()
-closeDBImpl VolatileDBEnv{..} = do
-    mbInternalState <- swapMVar _dbInternalState VolatileDbClosed
-    case mbInternalState of
-      VolatileDbClosed -> return ()
-      VolatileDbOpen InternalState{..} ->
-        wrapFsError hasFsErr _dbErr $ hClose _currentWriteHandle
+closeDBImpl VolatileDBEnv{..} =
+    bracket (takeMVar _dbInternalState)
+            (\_ -> putMVar _dbInternalState VolatileDbClosed)
+            (\_ -> unsafeReleaseAll _registry)
```

Comment:
I wasn't very sure about how thread safe `unsafeReleaseAll` is. This is why I always call it with the InternalState lock.
But I think you're right, this is not necessary.

Reply:
Hmm now that I think about it again, it may be possible that after swapping the Mvar to VolatileDBClosed and before cleaning the registry, another thread may reopen the db and allocate new resources, which should not be cleaned.

Before the registry, this was not possible, since we only closed the "old" handles, but with the registry we have scope to the new handles.

Reply:
> Hmm now that I think about it again, it's possible that after swapping the Mvar to VolatileDBClosed and before cleaning the registry, another thread may reopen the db and allocate new resources, which should not be cleaned.

Although it won't happen in practice, as we never reopen a VolatileDB (but we do in the tests, right?), you are technically correct. So you can leave the `bracket`.